backend/src/event/event.service.ts

```diff
@@ -0,0 +1,405 @@
+    private readonly venueService: VenueService
+  ) {}
+
+  public async createEvent(dto: EventCreateDTO, user: User): Promise<EventDTO> {
+    if (dto.status === StatusEnum.RESERVED) {
+      throw new BadRequestException(
+        'Cannot have RESERVED status on event creation'
+      );
+    }
+
+    const validationErrorDTO = await this.verifyEventCreation(dto);
+
+    if (validationErrorDTO.hasErrors) {
+      throw new BadRequestException(validationErrorDTO, 'Error creating event');
+    }
+
+    const event = await this.prismaService.event.create({
+      data: {
+        title: dto.title,
+        description: dto.description,
+        status: dto.status,
+        startTime: dto.startTime,
+        endTime: dto.endTime,
+        contactPerson: dto.contactPerson,
+        contactNumber: dto.contactNumber,
+        approvedBy: dto.approvedBy,
+        viewAccess: dto.viewAccess,
+        type: dto.type,
+        additionalNotes: dto.additionalNotes,
+        organizer: {
+          connect: { id: dto.organizerId },
+        },
+        encodedBy: { connect: { id: user.id } },
+        equipments: {
+          create: dto.equipmentIds.map((id) => ({
+            equipmentId: id,
+          })),
+        },
+        venues: {
+          create: dto.venueIds.map((id) => ({
+            venuesId: id,
+          })),
+        },
+      },
+      include: {
+        encodedBy: true,
+        organizer: true,
+        equipments: { include: { equipment: true } },
+        venues: { include: { venue: true } },
+      },
+    });
+
+    await this.addSchedules(dto);
+
+    return EventService.convertToDTO(event);
+  }
+
+  public async getAllEvents(): Promise<EventDTO[]> {
+    const events = await this.prismaService.event.findMany({
+      include: {
+        encodedBy: true,
+        organizer: true,
+        equipments: { include: { equipment: true } },
+        venues: { include: { venue: true } },
+      },
+    });
+
+    return events.map((e) => EventService.convertToDTO(e));
+  }
+
+  public async getEventById(id: string): Promise<EventDTO> {
+    try {
+      const event = await this.prismaService.event.findUniqueOrThrow({
+        where: { id },
+        include: {
+          encodedBy: true,
+          organizer: true,
+          equipments: { include: { equipment: true } },
+          venues: { include: { venue: true } },
+        },
+      });
+
+      return EventService.convertToDTO(event);
+    } catch (e) {
+      if (e instanceof NotFoundError) {
+        throw new NotFoundException();
+      }
+    }
+  }
+
+  public async updateEvent(
+    id: string,
+    dto: EventCreateDTO,
+    user: User
+  ): Promise<EventDTO> {
+    await this.deleteSchedules(id);
+
+    try {
+      await this.prismaService.equipmentsOnEvents.deleteMany({
+        where: { eventId: id },
+      });
+      await this.prismaService.venuesOnEvents.deleteMany({
+        where: { eventId: id },
+      });
```

Comment:
not sure if this is the best way to do this but i first needed to delete the existing equipment/venue relations before updating because otherwise, prisma complains that these relations must exist